mongo/mongo_test.go

```diff
@@ -413,7 +421,7 @@ func TestMongoHelpers(t *testing.T) {
 				arr, hasOutputStage, err := transformAggregatePipeline(bson.NewRegistryBuilder().Build(), tc.pipeline)
 				assert.Equal(t, tc.hasOutputStage, hasOutputStage, "expected hasOutputStage %v, got %v",
 					tc.hasOutputStage, hasOutputStage)
-				assert.Equal(t, tc.err, err, "expected error %v, got %v", tc.err, err)
+				assert.True(t, compareErrors(tc.err, err), "expected error %v, got %v", tc.err, err)
```

Comment:
How is `assert.Equal` different than `compareErrors` for checking equality of errors?

Reply:
Good question. `assert.Equal` does a deep comparison of the error structs which fails as the actual error is quite complex. Something like:

```
Error:      	Not equal:
        	            	expected: mongo.MarshalError{Value:primitive.D{}, Err:(*errors.errorString)(0x14000099d40)}
        	            	actual  : mongo.MarshalError{Value:primitive.D{primitive.E{Key:"hello", Value:(func())(0x1045477a0)}}, Err:bsoncodec.ErrNoEncoder{Type:(*reflect.rtype)(0x104719760)}}

        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1,7 +1,23 @@
        	            	 (mongo.MarshalError) {
        	            	- Value: (primitive.D) {
        	            	+ Value: (primitive.D) (len=1) {
        	            	+  (primitive.E) {
        	            	+   Key: (string) (len=5) "hello",
        	            	+   Value: (func()) 0x1045477a0
        	            	+  }
        	            	  },
        	            	- Err: (*errors.errorString)({
        	            	-  s: (string) (len=27) "no encoder found for func()"
        	            	- })
        	            	+ Err: (bsoncodec.ErrNoEncoder) {
        	            	+  Type: (*reflect.rtype)({
        	            	+   size: (uintptr) 0x8,
        	            	+   ptrdata: (uintptr) 0x8,
        	            	+   hash: (uint32) 2159672062,
        	            	+   tflag: (reflect.tflag) 2,
        	            	+   align: (uint8) 8,
        	            	+   fieldAlign: (uint8) 8,
        	            	+   kind: (uint8) 51,
        	            	+   equal: (func(unsafe.Pointer, unsafe.Pointer) bool) <nil>,
        	            	+   gcdata: (*uint8)(1),
        	            	+   str: (reflect.nameOff) 12938,
        	            	+   ptrToThis: (reflect.typeOff) 0
        	            	+  })
        	            	+ }
```
I used `compareErrors` to only assert the errors' strings matched.

Reply:
Ah, wow, that's quite different. You could consider using `assert.EqualError`, which seems to do something very similar to what `compareErrors` does. Otherwise, this looks good as is.

Reply:
Done. Using `assert.EqualError` here requires extra logic:
```
if tc.err != nil {
           assert.NotNil(t, err)
           assert.EqualError(t, err, tc.err.Error())
} else {
           assert.Nil(t, err)
}
```
Seems preferable to another function, though.